nav2_amcl/src/amcl_node.cpp

```diff
@@ -20,440 +20,219 @@
+  std::string tf_error;
 
-  tmp_pos = parameters_client->get_parameter("initial_pose_x", init_pose_[0]);
+  RCLCPP_INFO(get_logger(), "Checking that transform thread is ready");
 
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[0] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose X position");
-  }
-
-  tmp_pos = parameters_client->get_parameter("initial_pose_y", init_pose_[1]);
-
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[1] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose Y position");
-  }
+  while (rclcpp::ok() &&
+    !tf_buffer_->canTransform(global_frame_id_, odom_frame_id_, tf2::TimePointZero,
+    tf2::durationFromSec(0.1), &tf_error))
+  {
+    if (last_error + nav2_util::duration_from_seconds(1.0) < rclcpp_node_->now()) {
+      RCLCPP_INFO(get_logger(), "Timed out waiting for transform from %s to %s"
+        " to become available, tf error: %s",
+        odom_frame_id_.c_str(), global_frame_id_.c_str(), tf_error.c_str());
+      last_error = rclcpp_node_->now();
+    }
 
-  tmp_pos = parameters_client->get_parameter("initial_pose_a", init_pose_[2]);
+    // The error string will accumulate and errors will typically be the same, so the last
+    // will do for the warning above. Reset the string here to avoid accumulation.
+    tf_error.clear();
 
-  if (!std::isnan(tmp_pos)) {
-    init_pose_[2] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial pose Yaw");
+    // Let the transforms populate, then retry
+    std::this_thread::sleep_for(10ms);
   }
+}
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_xx", init_cov_[0]);
+nav2_lifecycle::CallbackReturn
+AmclNode::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
 
-  if (!std::isnan(tmp_pos)) {
-    init_cov_[0] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial covariance XX");
-  }
+  // Lifecycle publishers must be explicitly activated
+  pose_pub_->on_activate();
+  particlecloud_pub_->on_activate();
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_yy", init_cov_[1]);
+  first_pose_sent_ = false;
 
-  if (!std::isnan(tmp_pos)) {
-    init_cov_[1] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial covariance YY");
-  }
+  // Keep track of whether we're in the active state. We won't
+  // process incoming callbacks until we are
+  active_ = true;
 
-  tmp_pos = parameters_client->get_parameter("initial_cov_aa", init_cov_[2]);
+  // Wait until the transform listener thread has had a chance to spin up and get the
+  // transforms that it needs
+  waitForTransforms();
 
-  if (!std::isnan(tmp_pos)) {
-    init_cov_[2] = tmp_pos;
-  } else {
-    RCLCPP_WARN(get_logger(), "ignoring NAN in initial covariance AA");
+  // Make sure we've output the first pose before continuing
+  while (!first_pose_sent_) {
+    std::this_thread::sleep_for(100ms);
```

Comment:
@mhpanah As I mentioned to Carl in another comment I purposely have not added timeouts yet as I want to know the exact state of each node without having it fall through a timeout. This has helped in debugging race conditions at startup. Now that the bringup is in good shape (there may be another issue or two that we can uncover, or course), we should discuss timeouts, system monitoring, error recovery, and so with a system perspective in mind. I'd rather not just start adding timeouts without considering the bigger picture and how we can improve reliability and robustness.

Reply:
Sure, that makes sense. If we can spot potential places during code reviews that might need a timeout or other system monitoring in future, maybe we can put a `comment` so that in future we don't miss them. What do you think?

Reply:
@mhpanah I had a look at the various modules to remind myself about the various occurances. At the lifecycle node level, these should all be in on_activate and will be easy to find and update. What will require a little more study is any loops/waits in the plugins. I'll go through all of it and update per our (future) discussion).